scalyr_agent/builtin_monitors/kubernetes_events_monitor.py

```diff
@@ -0,0 +1,478 @@
+
+    def _get_oldest_node( self, nodes, ignore_master ):
+        """
+            Takes a list of nodes returned from querying the k8s api, and returns the name
+            of the node with the oldest creationTimestamp or None if nodes is empty or doesn't
+            contain Node objects.
+
+            if `ignore_master` is true, then any nodes with the label `node-role.kubernetes.io/master` are ignored.
+        """
+
+        oldest_time = datetime.datetime.utcnow()
+        oldest = None
+
+        # loop over all nodes and find the one with the oldest create time
+        for node in nodes:
+            metadata = node.get( 'metadata', {} )
+            create_time = metadata.get( 'creationTimestamp', None )
+            if create_time is None:
+                continue
+
+            name = metadata.get( 'name', None )
+            if name is None:
+                continue
+
+            # skip the master node if necessary
+            labels = metadata.get( 'labels', {} )
+            if ignore_master and 'node-role.kubernetes.io/master' in labels:
+                continue
+
+            #convert to datetime
+            create_time = scalyr_util.rfc3339_to_datetime( create_time )
+
+            # if we are older than the previous oldest datetime, then update the oldest time
+            if create_time is not None and create_time < oldest_time:
+                oldest_time = create_time
+                oldest = name
+
+        return oldest
+
+    def _check_nodes_for_leader( self, k8s, query_fields='' ):
+        """
+            Checks all nodes on the cluster to see which one has the oldest creationTime
+            If `ignore_master` is True then any node with the label `node-role.kubernetes.io/master`
+            is ignored.
+        """
+        response = k8s.query_api( '/api/v1/nodes%s' % query_fields )
+        nodes = response.get( 'items', [] )
+
+        # sort the list by node name to ensure consistent processing order
+        # if two nodes have exactly the same creation time
+        nodes.sort( key=lambda n: n.get( 'metadata', {} ).get( 'name', '' ) )
+
+        # the leader is the longest running node
+        return self._get_oldest_node( nodes, self._ignore_master )
+
+    def _get_current_leader( self, k8s ):
+        """
+            Queries the kubernetes api to see which node is the current leader node.
+
+            If there is currently a leader node, then only that node is polled to see if it still
+            exists.
+
+            Check to see if the leader is specified as a node label.
+            If not, then if there is not a current leader node, or if the current leader node no longer exists
+            then query all nodes for the leader
+        """
+
+        try:
+            leader = None
+            # first check to see if the leader node is specified as a label
+            # we do this first in case the label has been moved to a different node but the old node is still
```

Comment:
In this model, you are allowing the label to be on multiple nodes and the leader election picks the oldest of all of those nodes, right?

The drawback is I believe you have to always query on that label to determine if the set of nodes with the label has changed.

Reply:
Yes, that's correct, but before doing that, I'm also sorting nodes by node name first in case there are two equally old nodes, in which case it will always return a consistent result.

For the second point, I'm not sure that's a drawback.  

Label filtering/selecting is an integral part of k8s.  According to [the docs](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#label-selectors) label selectors are "the core grouping primitive in Kubernetes" so I would imagine this kind of query is going to be acceptably performant - even for large clusters.

I imagine the nodes that have this label set will be limited to a small handful.  The smallest case 1, but probably 2-3 just for fallback purposes.

When this function is called the `leader_check_interval` has expired.  This means we need to be hitting the api anyway firstly to determine if the `current_leader` is still alive, and if not, hitting it again to query *all the nodes*.

So instead of hitting the api to check if a single node is alive, we instead first query to get a short list of alive nodes that have this label attached.

In a best case scenario we are doing the same amount of calls to the API (because if we get nodes back, we don't have to check the alive status of the previously current node, or query the entire list of nodes in the cluster).

In the worse case scenario (no labels set and the current node is dead) we are doing one more API call (query on labels - which fails, query on current node - which fails, then query on all nodes), however a worst case scenario is predicated on configuring the monitor to use labels (which is off by default) and then not setting any labels.

The last possibility is the case where the previous leader node goes down, then assuming labels have been set on a currently running node, querying by this label is likely going to be more efficient, because it avoids having to then query and process all the nodes.

Reply:
Ok, we should make sure we call this out in the documentation that the number of nodes with the `could be leader` label should be small.  Also, let's rename that label.  `events_leader=true` makes it look like it is the event leader, where it could just be a candidate.  How about `events_leader_candidate=true`.  Kind of long, but not sure which if we really want to remove `events` here.

Reply:
Done.